runtime/eventring.c

```diff
@@ -0,0 +1,543 @@
+                               caml_domain_state **participating_domains) {
+  caml_global_barrier();
+  if (participating_domains[0] == domain_state) {
+#ifdef _WIN32
+    UnmapViewOfFile(current_metadata);
+    CloseHandle(ring_file_handle);
+    CloseHandle(ring_handle);
+#else
+    munmap(current_metadata, current_ring_total_size);
+
+    int remove_file = *(int*)data;
+
+    if( remove_file ) {
+      unlink(current_ring_loc);
+    }
+#endif
+
+    caml_stat_free(current_ring_loc);
+    current_metadata = NULL;
+
+    atomic_store_rel(&eventring_enabled, 0);
+  }
+  caml_global_barrier();
+}
+
+void caml_eventring_post_fork() {
+  /* We are after a call to fork (which can only happen when there is a single
+     domain) and no mutator code that can spawn a new domain can have run yet.
+     Let's be double sure. */
+  CAMLassert(caml_domain_alone());
+
+  if (atomic_load_acq(&eventring_enabled)) {
+    /* unmmap eventrings from the parent but don't remove the files */
+    int remove_file = 0;
+    caml_try_run_on_all_domains(&teardown_eventring, &remove_file, NULL);
+
+    /* We still have the path and ring size from our parent */
+    caml_eventring_start();
+  }
+}
+
+/* Return the current location for the ring buffers of this process. This is
+  used in the consumer to read the ring buffers of the current process */
+char_os* caml_eventring_current_location() {
+  if( atomic_load_acq(&eventring_enabled) ) {
+    return current_ring_loc;
+  } else {
+    return NULL;
+  }
+}
+
+/* Write a lifecycle event and then trigger a stop the world to tear down the
+  ring buffers */
+void caml_eventring_destroy() {
+  if (atomic_load_acq(&eventring_enabled)) {
+    write_to_ring(EV_RUNTIME, EV_LIFECYCLE, EV_RING_STOP, 0, NULL, 0);
+
+    /* clean up eventring when we exit. */
+    int remove_file = 1;
+    caml_try_run_on_all_domains(&teardown_eventring, &remove_file, NULL);
+  }
+}
+
+/* Create the ring buffers. This must only be called from a stop the world
+  pause. */
+static void
+create_and_start_ring_buffers(caml_domain_state *domain_state, void *data,
+                              int num_participating,
+                              caml_domain_state **participating_domains) {
+  /* Everyone must be stopped for starting and stopping eventring */
+  caml_global_barrier();
+
+  /* Only do this on one domain */
+  if (participating_domains[0] == domain_state) {
+    /* Don't initialise eventring twice */
+    if (!atomic_load_acq(&eventring_enabled)) {
+      int ret, ring_headers_length;
+#ifdef _WIN32
+      DWORD pid = GetCurrentProcessId();
+#else
+      int ring_fd;
+      long int pid = getpid();
+#endif
+
+      current_ring_loc = caml_stat_alloc(EVENTRING_MAX_MSG_LENGTH);
+
+      if (eventring_path) {
+        char* path_u8 = caml_stat_strdup_of_os(eventring_path);
+        snprintf_os(current_ring_loc, EVENTRING_MAX_MSG_LENGTH,
+                    T("%s/%ld.eventring"), path_u8, pid);
+        caml_stat_free(path_u8);
```

Comment:
Do we really need both `caml_stat_strdup_of_os` and then `snprintf_os`? I thought it was one or the other. (cc @dra27 or @nojb)

Reply:
(The `if (eventlog_filename)` test in eventlog.c looks very close to this code, but without the `caml_stat_strdup_of_os` dance, so it's much simpler. I hope that the simpler eventlog code is correct...)